src/appleseed/renderer/modeling/camera/fisheyelenscamera.cpp

```diff
@@ -0,0 +1,446 @@
+            m_pixel_area = m_film_dimensions[0] * m_film_dimensions[1] / pixel_count;
+
+            const string projection_type = m_params.get_required<string>("projection_type", "equisolid_angle");
+
+            if (projection_type == "equisolid_angle")
+                m_projection_type = EquisolidAngle;
+            else
+            {
+                RENDERER_LOG_ERROR(
+                    "invalid value \"%s\" for parameter \"projection_type\", "
+                    "using default value \"equisolid_angle\".",
+                    projection_type.c_str());
+                m_projection_type = EquisolidAngle;
+            }
+
+            return true;
+        }
+
+        void transform_ray_dir(Vector3d& dir) const
+        {
+            // Transform the direction vector to be in the xz plane.
+            const Transformd rot = Transformd(Matrix4d::make_rotation_z(-atan2 (dir[1], dir[0])));
+            dir = rot.vector_to_parent(dir);
+
+            if (m_projection_type == EquisolidAngle) 
+            {
+                // Tangent of angle between vector(on xz plane) and z+ axis.
+                const double tangent_xz = dir[0]/dir[2];
+                double theta1 = atan(tangent_xz);
+                double theta2 = 2 * asin(tangent_xz / 2.0);
+
+                const Transformd rot2 = Transformd(Matrix4d::make_rotation_y(theta2 - theta1));
+
+                dir = rot2.vector_to_parent(dir);
+                dir = rot.vector_to_local(dir);
+            }
+        }
+
+        void spawn_ray(
+            SamplingContext&    sampling_context,
+            const Dual2d&       ndc,
+            ShadingRay&         ray) const override
+        {
+
+            // Initialize the ray.
+            initialize_ray(sampling_context, ray);
+
+            // Retrieve the camera transform.
+            Transformd scratch;
+            const Transformd& transform =
+                m_transform_sequence.evaluate(ray.m_time.m_absolute, scratch);
+
+            // Compute ray origin and direction.
+            ray.m_org = transform.get_local_to_parent().extract_translation();
+            ray.m_dir = normalize(transform.vector_to_parent(-ndc_to_camera(ndc.get_value())));
+
+            transform_ray_dir(ray.m_dir);
+
+            // Compute ray derivatives.
+            if (ndc.has_derivatives())
+            {
+                const Vector2d px(ndc.get_value() + ndc.get_dx());
+                const Vector2d py(ndc.get_value() + ndc.get_dy());
+                ray.m_rx.m_org = ray.m_org;
+                ray.m_ry.m_org = ray.m_org;
+                ray.m_rx.m_dir = normalize(transform.vector_to_parent(-ndc_to_camera(px)));
+                transform_ray_dir(ray.m_rx.m_dir);
+                ray.m_ry.m_dir = normalize(transform.vector_to_parent(-ndc_to_camera(py)));
+                transform_ray_dir(ray.m_ry.m_dir);
+                ray.m_has_differentials = true;
+            }
+        }
+
+        bool connect_vertex(
```

Comment:
I worked this file based on `pinholecamera.cpp` copy, so it is still same with it. I'll take a look in soon.